lib/generator.ml

```diff
@@ -301,27 +332,33 @@ let make_upgraders ?(impl_kind = Config.Native) ?output_prefix = function
         | Error e ->
           Error e)
     in
-    let* versions = get_versions [] files in
-    let file_versions = List.sort ~cmp:Int.compare versions in
+    let* desc_file_versions = get_versions [] files in
+    let file_versions = List.rev desc_file_versions in
     let recreate_path version =
-      [%string "$folder/$(input_prefix)_%i$(version).$atd_extension"]
+      [%string "$folder/$(input_prefix)_$(version).$atd_extension"]
     in
     let prefix = Option.value ~default:input_prefix output_prefix in
-    let rec make_upgraders ~version_pairs ~main_type ~upgraders = function
+    let rec make_upgraders ~desc_file_versions ~version_pairs ~upgraders
+      = function
       | [] ->
-        assert false
+        failwith "received 0 ATD files as input in make_upgraders"
       | [ newest_version ] ->
-        let main_type =
-          match main_type with
-          | Some main_type ->
-            main_type
-          | None ->
-            let _sorted_items, _type_map, main_type, _main_type_param_count =
-              Upgrader.load_sort_map (recreate_path newest_version)
-            in
-            main_type
+        let* ( _sorted_items
+             , _type_map
+             , main_type
+             , _main_type_param_count
+             , latest_version )
+          =
```

Comment:
This generally points in the direction of creating separate functions to just return those few fields specifically. The exception being if you need to compute sorted_items/type_map/main_type_param_count anyway.

Reply:
yeah I know it looks ugly, but in practice for this case, the separate function would be pretty similar so I think it makes sense to keep it that way.